<a href="https://colab.research.google.com/github/chaitu713/data-analysis-with-python/blob/master/8.numpy-array-operations.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# A Beginner's Guide to Linear Algebra with NumPy 


NumPy is a python package for Scientific Computing. It provides high 
performance multidimensional array objects. It has functions for working in a domain of Linear Algebra, Fourier Transform and Matrices. NumPy was created in 2005 by Travis Oliphant and it is an open source project. In this notebook, let's go through the some of the NumPy linear algebra functions. 

Functions used in this notebook are:
- numpy.vdot
- numpy.inner
- numpy.linalg.det
- numpy.linalg.solve
- numpy.linalg.inv

Let's begin by importing Numpy and listing out the functions covered in this notebook.

In [1]:
import numpy as np

In [2]:
# List of functions explained 
function1 = np.vdot
function2 = np.inner
function3 = np.linalg.det
function4 = np.linalg.solve
function5 = np.linalg.inv

## Function 1 - np.vdot

- This function is used to calculate the dot product of two vectors. It can be defined as the sum of product of corresponding elements of n-dimensional arrays.

In [3]:
# Example 1 - working
arr1 = np.array([[23,21],[10,25]])
arr2 = np.array([[26,12],[20,11]])

result = np.vdot(arr1, arr2)

result

1325

np.vdot(arr1, arr2) - 23 * 26 + 21 * 12 + 10 * 20 + 25 * 11 = 1325

In [4]:
# Example 2 - working
arr1 = np.array([np.arange(2,8,3),np.arange(3,9,3)])
arr2 = np.array([np.arange(4,12,4),np.arange(5,15,5)])
result = np.vdot(arr1, arr2)
print("Array1\n", arr1)
print("")
print("Array2\n", arr2)
print("")
print("Result of vdot operation is ",result)

Array1
 [[2 5]
 [3 6]]

Array2
 [[ 4  8]
 [ 5 10]]

Result of vdot operation is  123


Generated two arrays using np.arange and have applied vdot on both the arrays.

np.vdot(arr1, arr2) - 2 * 3 + 5 * 8 + 3 * 5 + 6 * 10 = 123

In [5]:
# Example 3 - breaking (to illustrate when it breaks)
arr1 = np.array([np.arange(2,8,3),np.arange(3,9,3)])
arr2 = np.array([np.linspace(2,8,3),np.linspace(3,9,3)])

print("Array1\n", arr1)
print("")
print("Array2\n", arr2)
result = np.vdot(arr1, arr2)
print(result)

Array1
 [[2 5]
 [3 6]]

Array2
 [[2. 5. 8.]
 [3. 6. 9.]]


ValueError: ignored

Unlike the above examples, this example doesn't work because the sizes of both the arrays are different.

## Function 2 - np.inner

This function returns the sum of the product of inner elements of the 1-D array. For n-D arrays, it returns the sum of product of the elements over the last axis.

In [7]:
# Example 1 - working
arr1 = np.array(np.arange(2,10,2))
arr2 = np.array(np.arange(3,15,3))
print("Array1\n", arr1)
print("")
print("Array2\n", arr2)
result = np.inner(arr1, arr2)
print("")
print("Result of np.inner is : ",result)

Array1
 [2 4 6 8]

Array2
 [ 3  6  9 12]

Result of np.inner is :  180


np.inner(arr1, arr2) - 2 * 3 + 4 * 6 + 6 * 9 + 8 * 12 = 180

In [8]:
# Example 2 - working
arr1 = np.array([np.arange(2,8,3),np.arange(3,9,3)])
arr2 = np.array([np.arange(4,12,4),np.arange(5,15,5)])
result = np.inner(arr1, arr2)
print("Array1\n", arr1)
print("")
print("Array2\n", arr2)
print("")
print("Result of np.inner operation is : ",result)

Array1
 [[2 5]
 [3 6]]

Array2
 [[ 4  8]
 [ 5 10]]

Result of np.inner operation is :  [[48 60]
 [60 75]]


np.inner(arr1, arr2) - 

    [[2 * 4 + 5 * 8   3 * 4 + 6 * 8]
     [2 * 5 + 5 * 10   3 * 5 + 6 * 10]]

Result of np.inner operation is :  
    [[48 60]
     [60 75]]

In [9]:
# Example 3 - breaking (to illustrate when it breaks)
arr1 = np.array(np.arange(2,10,2))
arr2 = np.array(np.arange(3,15,2))
print("Array1\n", arr1)
print("")
print("Array2\n", arr2)
result = np.inner(arr1, arr2)
print("")
print("Result of np.inner is : ",result)

Array1
 [2 4 6 8]

Array2
 [ 3  5  7  9 11 13]


ValueError: ignored

Unlike the first 2 examples, this example doesn't work because the sizes of two arrays are different

## Function 3 - np.linalg.det

The determnant of a matrix can be calculated using diagonal elements. The det of 2 * 2 matrix of example 
            [[A B]
             [C D]] 
is AD - BC

In [10]:
# Example 1 - working
arr1 = np.array([np.arange(2,8,3),np.arange(3,9,3)])
print("Array1\n", arr1)
result = np.linalg.det(arr1)
print("")
print("Result of det operation is : ",result)

Array1
 [[2 5]
 [3 6]]

Result of det operation is :  -3.0000000000000004


Determinant of 2 * 2 array is AD - BC

In [11]:
# Example 2 - working
arr1 = np.array([np.arange(6,12,2),np.arange(5,10,2), np.arange(7,14,3)])
print("Array1\n", arr1)
result = np.linalg.det(arr1)
print("")
print("Result of det operation is : ",result)

Array1
 [[ 6  8 10]
 [ 5  7  9]
 [ 7 10 13]]

Result of det operation is :  0.0


Determinant of 3 * 3 array is 

Let array be 
[[ A  B C]
 [ D  E  F]
 [ G H I]]
  
 The det of this 3 * 3 array is A * (EI - FH) - B * (DI - FG) + C * (DH - EG)

In [12]:
# Example 3 - breaking (to illustrate when it breaks)
arr1 = np.array([np.arange(6,12,2),np.arange(5,10,2)])
print("Array1\n", arr1)
result = np.linalg.det(arr1)
print("")
print("Result of det operation is : ",result)

Array1
 [[ 6  8 10]
 [ 5  7  9]]


LinAlgError: ignored

Determinant operation only works for square matrices i.e, 2 * 2, 3 * 3, 4 * 4, and so on

Some closing comments about when to use this function.

## Function 4 - np.linalg.solve

This function is used to solve a quadratic equation where values can be given in form of a matrix.

We will solve a system of linear equations:
    3X + 5Y + 7Z = 11
    2X + 4Y + 8Z = 4
    4X + Y + Z = 3

In [13]:
# Example 1 - working
arr1 = np.array([[3,5,7], [2,4,8], [4,1,1]])
arr2 = np.array([11,4,3])
print("Array1\n", arr1)
print("")
print("Array2\n", arr2)
result = np.linalg.solve(arr1, arr2)
print("")
print("Result of np.linalg.solve is : ",result)

Array1
 [[3 5 7]
 [2 4 8]
 [4 1 1]]

Array2
 [11  4  3]

Result of np.linalg.solve is :  [ 0.  5. -2.]


We will solve a system of linear equations:
    3X + 5Y = 11
    2X + 4Y = 4

In [14]:
# Example 2 - working
arr1 = np.array([[3,5],[2,4]])
arr2 = np.array([11,4])
print("Array1\n", arr1)
print("")
print("Array2\n", arr2)
result = np.linalg.solve(arr1, arr2)
print("")
print("Result of np.linalg.solve is : ",result)

Array1
 [[3 5]
 [2 4]]

Array2
 [11  4]

Result of np.linalg.solve is :  [12. -5.]


We will solve a system of linear equations:
    3X + 5Y + 7Z = 11
    2X + 4Y + 8Z = 4
 
 But, np.linalg.solve works with square matrices

In [15]:
# Example 3 - breaking (to illustrate when it breaks)
arr1 = np.array([[3,5,7], [2,4,8]])
arr2 = np.array([11,4])
print("Array1\n", arr1)
print("")
print("Array2\n", arr2)
result = np.linalg.solve(arr1, arr2)
print("")
print("Result of np.linalg.solve is : ",result)

Array1
 [[3 5 7]
 [2 4 8]]

Array2
 [11  4]


LinAlgError: ignored

## Function 5 - np.linalg.inv

This function is used to calculate the multiplicative inverse of a matrix.

In [17]:
# Example 1 - working
arr1 = np.array([[1,3],[4,5]])
print("Array1\n", arr1)
result = np.linalg.inv(arr1)
print("")
print("Result of np.linalg.inv is : ",result)

Array1
 [[1 3]
 [4 5]]

Result of np.linalg.inv is :  [[-0.71428571  0.42857143]
 [ 0.57142857 -0.14285714]]


Multiplicative inverse can of a 2 * 2 matrix which is of the form 
    [[a b]
     [c d]]
 is (1/ad-bc)[[d -b]
              [-c a]]

In [18]:
# Example 2 - working
arr1 = np.array([np.arange(2,8,3),np.arange(3,9,4)])
print("Array1\n", arr1)
result = np.linalg.inv(arr1)
print("")
print("Result of np.linalg.inv is : ",result)

Array1
 [[2 5]
 [3 7]]

Result of np.linalg.inv is :  [[-7.  5.]
 [ 3. -2.]]


Multiplicative inverse can of a 2 * 2 matrix which is of the form 
    [[a b]
     [c d]]
 is (1/ad-bc)[[d -b]
              [-c a]]

In [19]:
# Example 3 - breaking (to illustrate when it breaks)
arr1 = np.array([[1,3],[4,5], [2,2]])
print("Array1\n", arr1)
result = np.linalg.inv(arr1)
print("")
print("Result of np.linalg.inv is : ",result)

Array1
 [[1 3]
 [4 5]
 [2 2]]


LinAlgError: ignored

Multiplicative Inverse can be only calculated on Square Matrices

## Conclusion

In this notebook, we have gone through some of the basic NumPy Linear Algebra functions.

For more you can refer : https://numpy.org/doc/stable/reference/routines.linalg.html

## Reference Links
Provide links to your references and other interesting articles about Numpy arrays:
* Numpy official tutorial : https://numpy.org/doc/stable/user/quickstart.html
* ...